### ISSUE TITLE
[Feature Request] Add users login and sign up system

### ISSUE THREAD

**Issue by askfriends** [libDrive/libDrive#3]
Add users login and sign up system with users rank and role.

**Comment by eliasbenb** [comment]
This would be easy to implement for regular deployments, but would cause issues with Heroku deployments. There is no way to access to the Heroku environment variables (which is where the users are stored).

The 2 main options I can see for doing this are:
- Creating a redis database to store the config using Heroku's addon. But Heroku only allows redis to be used by Heroku premium users or users who linked their credit/debit card to their Heroku account
- Storing the config inside of google drive, so the config can be updated and saved permanently, but this might be complicated and would require changing the current config workflow.

I might add this in the future, but not soon.

Thanks for the suggestion!

**Comment by askfriends** [comment]
> This would be easy to implement for regular deployments, but would cause issues with Heroku deployments. There is no way to access to the Heroku environment variables (which is where the users are stored).
> 
> The 2 main options I can see for doing this are:
> 
> * Creating a redis database to store the config using Heroku's addon. But Heroku only allows redis to be used by Heroku premium users or users who linked their credit/debit card to their Heroku account
> * Storing the config inside of google drive, so the config can be updated and saved permanently, but this might be complicated and would require changing the current config workflow.
> 
> I might add this in the future, but not soon.
> 
> Thanks for the suggestion!

i think you can implement that by using third party free database website i dont remember its name or url, but will write here when i got that

**Comment by aditya-shri** [comment]
> > This would be easy to implement for regular deployments, but would cause issues with Heroku deployments. There is no way to access to the Heroku environment variables (which is where the users are stored).
> > The 2 main options I can see for doing this are:
> > 
> > * Creating a redis database to store the config using Heroku's addon. But Heroku only allows redis to be used by Heroku premium users or users who linked their credit/debit card to their Heroku account
> > * Storing the config inside of google drive, so the config can be updated and saved permanently, but this might be complicated and would require changing the current config workflow.
> > 
> > I might add this in the future, but not soon.
> > Thanks for the suggestion!
> 
> i think you can implement that by using third party free database website i dont remember its name or url, but will write here when i got that

you might be thinking of mongodb.
It provides free to use database service.

**Comment by askfriends** [comment]
> > > This would be easy to implement for regular deployments, but would cause issues with Heroku deployments. There is no way to access to the Heroku environment variables (which is where the users are stored).
> > > The 2 main options I can see for doing this are:
> > > 
> > > * Creating a redis database to store the config using Heroku's addon. But Heroku only allows redis to be used by Heroku premium users or users who linked their credit/debit card to their Heroku account
> > > * Storing the config inside of google drive, so the config can be updated and saved permanently, but this might be complicated and would require changing the current config workflow.
> > > 
> > > I might add this in the future, but not soon.
> > > Thanks for the suggestion!
> > 
> > 
> > i think you can implement that by using third party free database website i dont remember its name or url, but will write here when i got that
> 
> you might be thinking of mongodb.
> It provides free to use database service.

yeah exactly.

**Comment by eliasbenb** [comment]
> you might be thinking of mongodb.

Sorry, you're right, I was thinking of MongoDB not redis. Either way, it still requires you to give Heroku your credit card details. Which is a solution I'm not willing to pursue.

I think the best route is to implement what I previously suggested

> * Storing the config inside of google drive, so the config can be updated and saved permanently, but this might be complicated and would require changing the current config workflow.

I've added the suggestion to my [to do list](https://github.com/libDrive/libDrive/tree/db871b9feaf6e6a22dc6c861feb64641858815e0#to-do-list)

**Comment by aditya-shri** [comment]
> > you might be thinking of mongodb.
> 
> Sorry, you're right, I was thinking of MongoDB not redis. Either way, it still requires you to give Heroku your credit card details. Which is a solution I'm not willing to pursue.
> 
> I think the best route is to implement what I previously suggested
> 
> > * Storing the config inside of google drive, so the config can be updated and saved permanently, but this might be complicated and would require changing the current config workflow.
> 
> I've added the suggestion to my [to do list](https://github.com/libDrive/libDrive/tree/db871b9feaf6e6a22dc6c861feb64641858815e0#to-do-list)

No you don't require credit card in heroku to use mongodb.
 You can directly access the database using the url frerom mongodb

**Comment by askfriends** [comment]
> > > you might be thinking of mongodb.
> > 
> > 
> > Sorry, you're right, I was thinking of MongoDB not redis. Either way, it still requires you to give Heroku your credit card details. Which is a solution I'm not willing to pursue.
> > I think the best route is to implement what I previously suggested
> > > * Storing the config inside of google drive, so the config can be updated and saved permanently, but this might be complicated and would require changing the current config workflow.
> > 
> > 
> > I've added the suggestion to my [to do list](https://github.com/libDrive/libDrive/tree/db871b9feaf6e6a22dc6c861feb64641858815e0#to-do-list)
> 
> No you don't require credit card in heroku to use mongodb.
> You can directly access the database using the url from mongodb

yes thats right, im using that with some other repo and its working fine, no need to add credit card info

**Comment by eliasbenb** [comment]
I just released [version 1.1.2](https://github.com/libDrive/libDrive/releases/tag/v1.2.0) which implements this feature. The reason it took so long was because Heroku users (most users) wouldn't be able to retain user sign ups after the app restarts. I made a workaround by saving the config on Google Drive (see [optional steps on the setup wiki](https://github.com/libDrive/libDrive/wiki/Setup#optional-steps), after that was completed I was able to work on the sign up page. Thank you for your interest in my project!

Reference commits:

https://github.com/libDrive/backend/commit/0d44c75f81ae30fbbef5405f2f3df164176b9ac0
https://github.com/libDrive/frontend/commit/bab7467383672541afb4320e6a7500c5113b868f
